src/Functions/concatWs.cpp

```diff
@@ -0,0 +1,176 @@
+            return result_type->createColumnConstWithDefaultValue(input_rows_count);
+
+        auto c_res = ColumnString::create();
+        c_res->reserve(input_rows_count);
+        const ColumnConst * col_sep = checkAndGetColumnConstStringOrFixedString(arguments[0].column.get());
+        if (!col_sep)
+            throw Exception(
+                ErrorCodes::ILLEGAL_COLUMN,
+                "Illegal column {} of first argument of function {}. Must be a constant String.",
+                arguments[0].column->getName(),
+                getName());
+        String sep_str = col_sep->getValue<String>();
+
+        const size_t num_exprs = arguments.size() - 1;
+        const size_t num_args = 2 * num_exprs - 1;
+
+        std::vector<const ColumnString::Chars *> data(num_args);
+        std::vector<const ColumnString::Offsets *> offsets(num_args);
+        std::vector<size_t> fixed_string_sizes(num_args);
+        std::vector<std::optional<String>> constant_strings(num_args);
+
+        bool has_column_string = false;
+        bool has_column_fixed_string = false;
+
+        for (size_t i = 0; i < num_exprs; ++i)
+        {
+            if (i != 0)
+                constant_strings[2 * i - 1] = sep_str;
+
+            const ColumnPtr & column = arguments[i + 1].column;
+            if (const ColumnString * col = checkAndGetColumn<ColumnString>(column.get()))
+            {
+                has_column_string = true;
+                data[2 * i] = &col->getChars();
+                offsets[2 * i] = &col->getOffsets();
+            }
+            else if (const ColumnFixedString * fixed_col = checkAndGetColumn<ColumnFixedString>(column.get()))
+            {
+                has_column_fixed_string = true;
+                data[2 * i] = &fixed_col->getChars();
+                fixed_string_sizes[2 * i] = fixed_col->getN();
+            }
+            else if (const ColumnConst * const_col = checkAndGetColumnConstStringOrFixedString(column.get()))
+                constant_strings[2 * i] = const_col->getValue<String>();
+            else
+                throw Exception(ErrorCodes::ILLEGAL_COLUMN,
+                    "Illegal column {} of argument of function {}", column->getName(), getName());
+        }
+
+        String pattern;
+        pattern.reserve(num_args * 2);
+        for (size_t i = 0; i < num_args; ++i)
+            pattern += "{}";
+
+        FormatImpl::formatExecute(
+            has_column_string,
+            has_column_fixed_string,
+            std::move(pattern),
+            data,
+            offsets,
+            fixed_string_sizes,
+            constant_strings,
+            c_res->getChars(),
+            c_res->getOffsets(),
+            input_rows_count);
+        return std::move(c_res);
+    }
+
+private:
+    ContextWeakPtr context;
+};
+
+struct NameConcatWs
+{
+    static constexpr auto name = "concatWs";
+};
+struct NameConcatWsAssumeInjective
+{
+    static constexpr auto name = "concatWsAssumeInjective";
```

Comment:
It is better to change those function names to more ClickHouse-like names and add a separate alias to be fully compatible with Spark. An example of this can be found here: https://github.com/ClickHouse/ClickHouse/pull/43313/files 
```suggestion
    static constexpr auto name = "concatWithDelimiter";
};
struct NameConcatWsAssumeInjective
{
    static constexpr auto name = "concatWithDelimiterAssumeInjective";
```